task_processing/plugins/mesos/execution_framework.py

```diff
@@ -430,6 +423,28 @@ def resourceOffers(self, driver, offers):
                     log.info("Suppressing offers, no more tasks to run.")
                 continue
 
+        with_maintenance_window = [
+            offer for offer in offers
+            if offer.unavailability
+        ]
+
+        for offer in with_maintenance_window:
+            start_time = offer.unavailability.start['nanoseconds']
+            completion_time = int(
+                (start_time + offer.unavailability.duration['nanoseconds'])
+                / 1000000000
+            )
+            now = int(time.time())
+            duration = completion_time - now
+            if not time.time() > completion_time:
+                self.blacklist_slave(
+                    agent_id=offer.agent_id.value,
+                    timeout=duration,
+                )
+
+        without_maintenance_window = [
+            offer for offer in offers if offer not in with_maintenance_window]
```

Comment:
maybe move `without_maintenance_window` up to `with_maintenance_window` and use `not offer.unavailability` as condition? would read better to me. (also webscale!)